components/eamxx/src/share/util/eamxx_time_interpolation.cpp

```diff
@@ -0,0 +1,343 @@
+ *        y* = w*y0 + (1-w)*y1
+ * where w = (t1-dt)/(t1-t0)
+ * Input:
+ *   time_in - A timestamp to interpolate onto.
+ * Output
+ *   A map of (field name) and (Field) pairs such that each field is the interpolated values and
+ *   the map makes it possible to quickly query individual outputs.
+ */
+std::map<std::string,Field> TimeInterpolation::perform_time_interpolation(const TimeStamp& time_in)
+{
+  // Declare the output map
+  std::map<std::string,Field> interpolated_fields;
+
+  // If data is handled by files we need to check that the timestamps are still relevant
+  if (m_file_data_triplets.size()>0) {
+    check_and_update_data(time_in);
+  }
+
+  // Gather weights for interpolation.  Note, timestamp differences are integers and we need a
+  // real defined weight.
+  const Real w_num = m_time1 - time_in;
+  const Real w_den = m_time1 - m_time0;
+  const Real weight = w_num/w_den;
+
+  // Cycle through all stored fields and conduct the time interpolation
+  for (auto name : m_field_names)
+  {
+    const auto& field0 = m_fm_time0->get_field(name);
+    const auto& field1 = m_fm_time1->get_field(name);
+    Field field_out    = field0.clone();
+    field_out.update(field1,1.0-weight,weight);
+    interpolated_fields.emplace(name,field_out);
+  }
+
+  // Return map of interpolated fields
+  return interpolated_fields;
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function which registers a field in the local field managers.
+ * Input:
+ *   field_in - Is a field with the appropriate dimensions and metadata to match the interpolation
+ * Output:
+ *   None
+ */
+void TimeInterpolation::add_field(const Field& field_in)
+{
+  // First check that we haven't already added a field with the same name.
+  const auto name = field_in.name();
+  EKAT_REQUIRE_MSG(!m_fm_time0->has_field(name) and !m_fm_time1->has_field(name),
+		  "Error!! TimeInterpolation:add_field, field + " << name << " has already been added." << "\n");
+
+  // Clone the field for each field manager to get all the metadata correct.
+  auto field0 = field_in.clone();
+  auto field1 = field_in.clone();
+  m_fm_time0->add_field(field0);
+  m_fm_time1->add_field(field1);
+  m_field_names.push_back(name);
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function to shift the data of a single field from time1 to time0
+ * Input:
+ *   name - The name of the field to be shifted.
+ */
+void TimeInterpolation::shift_data(const std::string& name)
+{
+        auto  field0 = m_fm_time0->get_field(name);
+  const auto& field1 = m_fm_time1->get_field(name);
+  field0.deep_copy(field1);
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function to shift all data from time1 to time0, update timestamp for time0
+ */
+void TimeInterpolation::shift_data()
```

Comment:
Can we remove the above method, and put its impl directly in here then? Will we ever need to shift data only for some field but not others? If not, there's no point in having 2 methods, I think.

Reply:
That's a good question.  Currently, a user is allowed to update field data on a field by field basis (if they want).  This means we would want to only "shift_data" for that field at that time.  I envisioned a use case where we want to deal with internal interpolation, for example for an implicit solve.  So we march individual fields forward in time to populate the interpolation bounds and this might have to be done on a field by field basis.  But maybe this is a case where we add this use case when we need it, rather than design with it in mind.

Reply:
I think it's safe to assume we update all of them. If we have two groups of fields, we can create two TimeInterpolation obj.